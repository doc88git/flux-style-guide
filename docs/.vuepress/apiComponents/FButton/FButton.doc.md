### Props

| Prop name     | Description | Type    | Values | Default   |
| ------------- | ----------- | ------- | ------ | --------- |
| outline       |             | boolean | -      |           |
| flat          |             | boolean | -      |           |
| small         |             | boolean | -      |           |
| bigger        |             | boolean | -      |           |
| dense         |             | boolean | -      |           |
| circle        |             | boolean | -      |           |
| label         |             | string  | -      |           |
| disabled      |             | boolean | -      |           |
| inverseColor  |             | boolean | -      |           |
| icon          |             | string  | -      | ''        |
| iconColor     |             | string  | -      | 'primary' |
| color         |             | string  | -      | ''        |
| radius        |             | boolean | -      | true      |
| textUppercase |             | boolean | -      | true      |
| textColor     |             | string  | -      |           |

### Events

| Event name | Type      | Description |
| ---------- | --------- | ----------- |
| click      | undefined |
| mouseover  | undefined |
| mouseleave | undefined |

### Slots

| Name    | Description | Bindings |
| ------- | ----------- | -------- |
| default |             |          |
