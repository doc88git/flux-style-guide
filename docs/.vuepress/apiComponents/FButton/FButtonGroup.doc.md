### Props

| Prop name     | Description | Type           | Values | Default   |
| ------------- | ----------- | -------------- | ------ | --------- |
| options       |             | array          | -      |           |
| outline       |             | boolean        | -      |           |
| tab           |             | boolean        | -      |           |
| default       |             | string\|number | -      |           |
| textUppercase |             | boolean        | -      | true      |
| size          |             | string         | -      | 'default' |

### Events

| Event name | Type      | Description |
| ---------- | --------- | ----------- |
| change     | undefined |
