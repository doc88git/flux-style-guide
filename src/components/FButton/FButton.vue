<template>
  <component
    :is="componentType"
    class="btn"
    type="button"
    :class="btnStyle"
    :disabled="disabled"
    v-on="$listeners"
    v-bind="$attrs"
  >
    <div class="btn__inner" :class="btnInnerCenter">
      <div class="btn__inner__icon" :class="btnCenter" v-if="icon">
        <f-icon :name="icon" />
      </div>
      <div v-if="hasContent" class="btn__inner__content">
        <slot> {{ label }} </slot>
      </div>
    </div>
  </component>
</template>

<script>
import { FIcon } from '../FIcon'

export default {
  name: 'FButton',
  components: {
    FIcon
  },
  props: {
    link: Boolean,
    outline: Boolean,
    flat: Boolean,
    small: Boolean,
    bigger: Boolean,
    dense: Boolean,
    label: String,
    disabled: Boolean,
    icon: String,
    color: {
      type: String,
      default: ''
    },
    radius: {
      type: Boolean,
      default: true
    },
    textColor: String
  },
  computed: {
    componentType() {
      return this.link ? 'a' : 'button'
    },
    hasDefaultSlot() {
      return !!this.$slots.default
    },
    hasContent () {
      return !!this.label || !!this.hasDefaultSlot
    },
    hasName() {
      return this.label === ''
    },
    hasBg() {
      return !this.flat && !this.outline
    },
    btnColors () {
      return {
        [`color--text--${this.textColor}`]: !!this.textColor,
        [`color--flat--${this.color}`]: !!this.color && this.flat,
        [`color--outline--${this.color}`]: !!this.color && this.outline,
        [`color--default--${this.color}`]: !!this.color && !this.flat && !this.outline,
      }
    },
    btnSize () {
      return {
        ['btn--small']: this.small,
        ['btn--bigger']: this.bigger
      }
    },
    btnBox () {
      return {
        ['btn--default']: !this.flat && !this.outline,
        ['btn--outline']: this.outline,
        ['btn--flat']: this.flat,
        ['btn--dense']: this.dense,
        ['btn--noradius']: !this.radius
      }
    },
    btnStyle() {
      return [{ ...this.btnColors, ...this.btnSize, ...this.btnBox }]
    },
    btnCenter() {
      return this.hasContent ? '' : ['btn__inner__icon--center']
    },
    btnInnerCenter() {
      return this.hasContent ? '' : ['btn__inner--center']
    },
  }
}
</script>

<style lang="scss">
.btn {
  display: inline-flex;
  align-items: center;

  text-align: center;
  padding: 0.5rem 0.75rem;
  height: fit-content;
  margin: 0.25rem;
  border-radius: 0.25rem;
  text-transform: uppercase;
  width: auto;
  border: none;
  cursor: pointer;

  &:hover {
    outline: 0;
    opacity: 0.7;
  }

  &:focus {
    outline: 0;
  }

  &:active {
    outline: 0;
  }

  &:disabled {
    opacity: 50%;
    cursor: default;
  }

  &--noradius {
    border-radius: 0px;
  }

  &--default {
    color: var(--color-white);
    padding-right: 1.25rem;
    padding-left: 1.25rem;
    background-color: var(--color-primary);
    border: 1px solid var(--color-primary);
  }

  &--small {
    text-align: center;
    padding: 0.25rem 1.25rem;

    * {
      font-size: var(--text-sm);
    }
  }

  &--bigger {
    font-size: 1rem;
    height: 4rem;
    padding-right: 1.25rem;
    padding-left: 1.25rem;
    margin: 0.25rem;
    width: auto;
  }

  &--bigger-x {
    padding: 0.5rem 1rem;
    margin: 0.25rem;
    height: 5rem;
    font-size: 1.25rem;
  }

  &--dense {
    padding: 0 0.5rem;
    margin: 0;
  }

  &--flat {
    background: transparent;
  }

  &--outline {
    color: var(--color-primary);
    border: 1px solid;
    border-radius: 0.25rem;
    padding-left: 1.25rem;
    padding-right: 1.25rem;
  }

  &__inner {
    display: flex;
    flex-wrap: nowrap;
    align-items: center;
    align-content: center;
    width: 100%;
    &--center {
      justify-content: center;
    }
    &__icon {
      height: 100%;
      display: flex;
      align-items: center;
      margin-right: 0.5rem;
      padding: 0.5rem 0;
      &--center {
        margin-right: 0;
      }
    }
    &__content {
      height: 100%;
      text-align: center;
      width: 100%;
      letter-spacing: 1px;
    }
  }
}
</style>
