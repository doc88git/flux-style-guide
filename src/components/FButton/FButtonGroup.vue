<template>
  <div class="f-button-group">
    <f-button
      v-for="opt in options"
      :key="opt.value"
      @click="change(opt.value)"
      :class="{
        ...classes,
        'f-button-group__tab--selected': tab && opt.value === selected
      }"
      v-bind="getAttrs(opt.value)"
    >
      {{ opt.label }}
    </f-button>
  </div>
</template>

<script>
import FButton from './FButton'

export default {
  name: 'f-button-group',
  components: {
    FButton
  },
  props: {
    options: {
      type: Array,
      required: true
    },
    outline: Boolean,
    tab: Boolean,
    default: [String, Number],
    size: {
      type: String,
      default: ''
    }
  },
  data: () => ({
    selected: null
  }),
  computed: {
    isDefault() {
      return !this.outline && !this.tab
    },
    classes() {
      return {
        'f-button-group__tab': this.tab
      }
    },
    btnOptions () {
      return {
        flat: this.tab,
        small: this.size === 'small',
        bigger: this.size === 'bigger'
      }
    }
  },
  methods: {
    change(value) {
      this.selected = value
      this.$emit('change', this.selected)
    },
    getAttrs(id) {
      return this.tab || this.outline && id === this.selected
        ? { ...this.btnOptions, outline: false }
        : this.isDefault
        ? { ...this.btnOptions, outline: id === this.selected }
        : { ...this.btnOptions, outline: true }
    },
    isSelected(id) {
      return id === this.selected
    }
  },
  mounted() {
    if (this.default) this.change(this.default)
  }
}
</script>

<style lang="scss" scoped>
.f-button-group {
  display: flex;
  &__tab {
    border-radius: 0;
    color: var(--color-gray);
    &--selected {
      color: var(--color-primary);
      border-bottom-width: 1px;
      border-color: var(--color-primary);
    }
  }
}
</style>
