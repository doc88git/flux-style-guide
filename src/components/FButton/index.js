import FButton from "./FButton";
import FButtonGroup from "./FButtonGroup";

export { FButton, FButtonGroup };
