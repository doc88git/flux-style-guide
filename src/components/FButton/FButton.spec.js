import { shallowMount } from '@vue/test-utils'
import FButton from './FButton'

const WRAPPER_PROPS = { label: 'Button' }
const WRAPPER_SLOTS = { slots: { default: 'Teste' } }

describe('FButton Tests', () => {
  let WRAPPER

  beforeEach(() => WRAPPER = shallowMount(FButton, { propsData: WRAPPER_PROPS }))

  test('Componente existe', () => {
    expect(WRAPPER.exists()).toBe(true)
  })

  test('Componente nÃ£o alterado', () => {
    expect(WRAPPER.html()).toMatchSnapshot()
  })

  describe('Teste de computeds', () => {
    describe('Testes da computed btnInnerCenter', () => {
      test('Computed btnCenter quanto label: "Button"', () => {
        expect(WRAPPER.vm.btnInnerCenter).toBe('')
      })

      test('Computed btnCenter quanto $slots.default: "Teste"', () => {
        const wrapperWithSlot = shallowMount(FButton, WRAPPER_SLOTS)
        expect(wrapperWithSlot.vm.btnInnerCenter).toBe('')
      })

      test('Computed btnCenter quanto label: ""', () => {
        WRAPPER.setProps({ label: '' })
        expect(WRAPPER.vm.btnInnerCenter).toStrictEqual(['btn__inner--center'])
      })
    })

    describe('Testes da computed btnCenter', () => {
      test('Computed btnCenter quanto label: "Button"', () => {
        expect(WRAPPER.vm.btnCenter).toBe('')
      })

      test('Computed btnCenter quanto $slots.default: "Teste"', () => {
        const wrapperWithSlot = shallowMount(FButton, WRAPPER_SLOTS)
        expect(wrapperWithSlot.vm.btnCenter).toBe('')
      })

      test('Computed btnCenter quanto label: ""', () => {
        WRAPPER.setProps({ label: '' })
        expect(WRAPPER.vm.btnCenter).toStrictEqual(['btn__inner__icon--center'])
      })
    })

    describe('Testes da computed hasDefaultSlot', () => {
      test('Computed hasDefaultSlot quando $slots.default: null', () => {
        expect(WRAPPER.vm.hasDefaultSlot).toBe(false)
      })

      test('Computed hasDefaultSlot quando $slots.default: "Teste"', () => {
        expect(WRAPPER.vm.hasDefaultSlot).toBe(false)
      })
    })

    describe('Testes da computed hasContent', () => {
      test('Computed hasContent quando label: "Button', () => {
        expect(WRAPPER.vm.hasContent).toBe(true)
      })

      test('Computed hasContent quando label: "" e $slots.default: null', () => {
        WRAPPER.setProps({ label: '' })
        expect(WRAPPER.vm.hasContent).toBe(false)
      })

      test('Computed hasContent quando label: "" e $slots.default: "Teste"', () => {
        const wrapperWithSlot = shallowMount(FButton, WRAPPER_SLOTS)
        expect(wrapperWithSlot.vm.hasContent).toBe(true)
      })
    })

    describe('Testes da computed componentType', () => {
      test('Computed componentType quando link: true', () => {
        WRAPPER.setProps({ link: true })
        expect(WRAPPER.vm.componentType).toBe('a')
      })

      test('Computed componentType quando link: false', () => {
        WRAPPER.setProps({ link: false })
        expect(WRAPPER.vm.componentType).toBe('button')
      })
    })

    describe('Testes da computed hasName', () => {
      test('Computed hasName quando label: "Button"', () => {
        expect(WRAPPER.vm.hasName).toBe(false)
      })

      test('Computed hasName quando label: ""', () => {
        WRAPPER.setProps({ label: '' })
        expect(WRAPPER.vm.hasName).toBe(true)
      })
    })

    describe('Testes da computed hasBg', () => {
      test('Computed hasBg quanto flat: false e outline: false', () => {
        WRAPPER.setProps({ flat: false, outline: false })
        expect(WRAPPER.vm.hasBg).toBe(true)
      })

      test('Computed hasBg quanto flat: true e outline: false', () => {
        WRAPPER.setProps({ flat: true, outline: false })
        expect(WRAPPER.vm.hasBg).toBe(false)
      })

      test('Computed hasBg quanto flat: false e outline: true', () => {
        WRAPPER.setProps({ flat: true, outline: false })
        expect(WRAPPER.vm.hasBg).toBe(false)
      })

      test('Computed hasBg quanto flat: false e outline: false', () => {
        WRAPPER.setProps({ flat: true, outline: false })
        expect(WRAPPER.vm.hasBg).toBe(false)
      })
    })

    describe('Testes da computed btnColors', () => {
      test('Computed btnColors quando textColor: "black"', () => {
        const objContains = { ['color--text--black']: true }
        WRAPPER.setProps({ textColor: 'black' })
        expect(WRAPPER.vm.btnColors).toEqual(expect.objectContaining(objContains))
      })

      test('Computed btnColors quando color: "secondary" e flat: true', () => {
        const objContains = { ['color--flat--secondary']: true }
        WRAPPER.setProps({ color: 'secondary', flat: true })
        expect(WRAPPER.vm.btnColors).toEqual(expect.objectContaining(objContains))
      })

      test('Computed btnColors quando color: "secondary" e outline: true', () => {
        const objContains = { ['color--outline--secondary']: true }
        WRAPPER.setProps({ color: 'secondary', outline: true })
        expect(WRAPPER.vm.btnColors).toEqual(expect.objectContaining(objContains))
      })

      test('Computed btnColors quando color: "secondary", outline: false e flat: false', () => {
        const objContains = { ['color--default--secondary']: true }
        WRAPPER.setProps({ color: 'secondary' })
        expect(WRAPPER.vm.btnColors).toEqual(expect.objectContaining(objContains))
      })
    })

    describe('Testes da computed btnSize', () => {
      test('Computed btnSize quando prop small: true', () => {
        const objContains = { ['btn--small']: true }

        WRAPPER.setProps({ small: true })
        expect(WRAPPER.vm.btnSize).toEqual(expect.objectContaining(objContains))
      })

      test('Computed btnSize quando prop bigger: true', () => {
        const objContains = { ['btn--bigger']: true }

        WRAPPER.setProps({ bigger: true })
        expect(WRAPPER.vm.btnSize).toEqual(expect.objectContaining(objContains))
      })
    })

    describe('Testes da computed btnBox', () => {
      test('btnStyle com prop radius: false', () => {
        const objContains = { ['btn--noradius']: true }

        WRAPPER.setProps({ radius: false})
        expect(WRAPPER.vm.btnBox).toEqual(expect.objectContaining(objContains))
      })

      test('btnStyle com prop dense: true', () => {
        const objContains = { ['btn--dense']: true }

        WRAPPER.setProps({ dense: true })
        expect(WRAPPER.vm.btnBox).toEqual(expect.objectContaining(objContains))
      })

      test('btnStyle com prop outline: true', () => {
        const objContains = { ['btn--outline']: true }

        WRAPPER.setProps({ outline: true })
        expect(WRAPPER.vm.btnBox).toEqual(expect.objectContaining(objContains))
      })

      test('btnStyle com prop flat: true', () => {
        const objContains = { ['btn--flat']: true }

        WRAPPER.setProps({ flat: true })
        expect(WRAPPER.vm.btnBox).toEqual(expect.objectContaining(objContains))
      })

      test('btnStyle com prop flat: false e outline: false', () => {
        const objContains = { ['btn--default']: true }

        WRAPPER.setProps({ outline: false, flat: false })
        expect(WRAPPER.vm.btnBox).toEqual(expect.objectContaining(objContains))
      })
    })
  })
})