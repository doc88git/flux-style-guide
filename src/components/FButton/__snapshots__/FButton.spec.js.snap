// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FButton Tests Componente n√£o alterado 1`] = `
"<button type=\\"button\\" class=\\"btn btn--default\\">
  <div class=\\"btn__inner\\">
    <!---->
    <div class=\\"btn__inner__content\\"> Button </div>
  </div>
</button>"
`;
